src/parallel.cpp

```diff
@@ -549,6 +549,10 @@ void HandleBlockMessageThread(CNode *pfrom, const string strCommand, const CBloc
             if (nDoS > 0)
                 dosMan.Misbehaving(pfrom, nDoS);
         }
+
+        // the current fork is bad due to this block so reset the best header to the best fully-validated block
+        // so we can download another fork of headers (and blocks).
+        pindexBestHeader = chainActive.Tip();
```

Comment:
is this making pindexBestHeader = chainactive  necessary?  Because we are now multithreaded, it is possible that while you were processing this block you received other headers already on a different chain (the good chain) and pindexBestHeaeder would already have more chainwork..now you're resetting it here to be less than what it should be ... Furthermore,   If the block fails validation, then pindexBestHeader will already be the chaintip, unless other headers have arrived...i don't think adding this is helps and actually adds a new problem.